src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -1187,6 +1192,25 @@ private async ValueTask<bool> ReadNextDataFrameAsync(HttpResponseMessage respons
         public void Trace(string message, [CallerMemberName] string? memberName = null) =>
             _connection.Trace(StreamId, message, memberName);
 
+        private void AbortStream()
+        {
+            // ToDo : locking??? we don't have any in H/3 stream, we don't have a sync root or anything.
```

Comment:
I was looking at H2Stream corresponding method:
https://github.com/dotnet/runtime/blob/ae0e5e8341b2adf172f585d839e518a58c1d898e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L1292-L1292
And the main difference is that we lack any locking mechanism here, which kind of worries me.

Reply:
TL;DR: if we check the flag and it's not EOF and on the next line when we call Abort the flag is changed already - it is not much better than just calling Abort without any checks, is it?

We actually had a bit of a discussion on using flags to symbolize EOFs in my first attempt on cancellation PR https://github.com/dotnet/runtime/pull/54334#discussion_r656402567

I guess we might want to put some proper locking here. as HTTP/2 has.. or just always call abort ant let quic stream/msquic handle the race? If we imagine Dispose being called synchronously after finishing reading the response -- I imagine graceful shutdown should be complete inside QuicStream by that time. So Abort will be just a no-op. But I may be wrong. Always calling Abort may lead to unexpected behavior in streaming scenarios 🤔 but I feel like this check without any locks will not be much better that always aborting.

@geoffkizer what do you think?

Reply:
I'd say that checking is still better then not checking at all. There's a potential for race, but in a happy path (dispose the response after you've finished the reading), this should work and not send any aborts.

Anyway, I was surprised that there's no locking mechanism at all. So my question is, do we use any other technique I'm missing? Or is this an oversight that just works because we're testing the right thing? Or are we just not thread-safe on the level of a response message thus we don't care here?